src/main/scala/util/HeterogeneousBag.scala

```diff
@@ -2,17 +2,17 @@
 
 package freechips.rocketchip.util
 
-import Chisel._
-import chisel3.Record
-import scala.collection.immutable.ListMap
+import chisel3._
+import chisel3.experimental.DataMirror
+import scala.collection.immutable.VectorMap
 
 final case class HeterogeneousBag[T <: Data](elts: Seq[T]) extends Record with collection.IndexedSeq[T] {
-  def apply(x: Int) = elts(x)
+  val elements = VectorMap(elts.zipWithIndex.map { case (n,i) => (i.toString, DataMirror.internal.chiselTypeClone(n)) }:_*)
+
+  def apply(x: Int): T = elements.values.toSeq(x)
```

Comment:
```suggestion
  def apply(x: Int): T = {
    val key = elements.keys(x)
    elements(key)
  }
```
Now that `elements` is a `VectorMap`, we can make `apply` actually `O(1)` (as is implied by extending `IndexedSeq`), whereas the current implementation is `O(n)`.